R/splitBy.R

```diff
@@ -0,0 +1,264 @@
+           signature = c("x"),
+           function(x, MARGIN = NULL, ...)
+               standardGeneric("unsplitBy"))
+
+#' @rdname splitBy
+#' @export
+setMethod("unsplitBy", signature = c(x = "list"),
+    function(x, MARGIN = NULL, ...){
+        ############################## INPUT CHECK #############################
+        # Check MARGIN
+        if( !(MARGIN == "row" || MARGIN == "col" || is.null(MARGIN) ||
+              (is.numeric(MARGIN) && (MARGIN == 1 || MARGIN == 2))) ){
+            stop("'MARGIN' must be 1, 2, or NULL",
+                 call. = FALSE)
+        }
+        ############################ INPUT CHECK END ###########################
+        # Can data be combined row-wise or column-wise
+        # If row_wise, then column names should be equal
+        row_wise <- all(sapply(x, FUN = function(obj){
+            identical( colnames(obj), colnames(x[[1]])) 
+        }))
+        # If col-wise, then rownames should be equal
+        col_wise <- all(sapply(x, FUN = function(obj){
+            identical( rownames(obj), rownames(x[[1]])) 
+        }))
+        
+        # Combine data
+        # If data can be combined in both directions, and MARGIN is not specified
+        if( row_wise && col_wise && is.null(MARGIN) ){
+            stop("Both rownames and colnames match. Specify with 'MARGIN' ",
+                 "whether to combine data column-wise or row-wise.",
+                 call. = FALSE)
+            # If data is combined column-wise
+        } else if( col_wise && (MARGIN == "col" || MARGIN == 2 || is.null(MARGIN)) ){
+            se <- do.call(cbind, x)
+            # If data is combined row-wise
+        } else if( row_wise && (MARGIN == "row" || MARGIN == 1 || is.null(MARGIN)) ){
+            se <- do.call(rbind, x)
+            # If MARGIN is specified but rownames do not match
+        } else if ( !col_wise && (MARGIN == "col" || MARGIN == 2) ){
+            stop("rownames should match when data is combined column-wise",
+                 call. = FALSE)
+            # If MARGIN is specified but colnames do not match
+        } else if ( !row_wise && (MARGIN == "row" || MARGIN == 1) ){
+            stop("colnames should match when data is combined row-wise",
+                 call. = FALSE)
+            # If rownames nor colnames do not match, data cannot be combined
+        } else{
+            stop("Data cannot be combined since rownames nor colnames do not match.",
+                 call. = FALSE)
+        }
+        return(se)
+    }
+)
+   
+#' @rdname splitBy
+#' @export
+setMethod("unsplitBy", signature = c(x = "SimpleList"),
+    function(x, MARGIN = NULL, ...){
+        # Convert into a list
+        x <- as.list(x)
+        unsplitBy(x, MARGIN, ...)
+    }
+)
+################################ HELP FUNCTIONS ################################
+# Split data in column-wise or row-wise based on grouping variable. 
+.split_by <- function(x, grouping, MARGIN, use_names = TRUE, ...){
```

Comment:
`splitBy` does not perform agglomeration, but `splitByRanks` does. I would suggest you use the example with `mergeRows` and implement them. Otherwise `splitBy` is just a fancy wrapper for `[`.

Having set that an option `skip_agglomerate` is certainly a valid addition, but the focus should be to get as close as possible to what `splitByRank` does. That has also been the expectation when @antagomir raised the issue

Reply:
`splitByRanks `
1. agglomerates the data at every rank (or one specific rank if rank is specified)
2. Returns a list of `SE` objects; each includes data that is agglomerated at certain rank (Kingdom, Phylum, ...)

Should `splitBy` do the same for columns? 
1. agglomerate the data at every variable of `colData`
2. Return a list of SE objects, each object represent each variable (in case of GlobalPatterns data: sampleType, Primer, ..)


Currently, `splitBy` returns a list of `SE` objects. Each object represent each group of specified variable.
For example, it could include Soil, Skin,... when SampleType variable is chosen

Reply:
I think we need 

1. splitting for samples by groups in colData (e.g. each list element has sample per one sample category i.e. sampleType)
2. splitting for features by groups in rowData (e.g. each list element has features per one feature category i.e. Phylum)
3. the aggregated version for multiple fields as you describe above.

Potentially 1-2 are easy to handle with some one-liner; the 3 is more akin to `splitByRanks`.

I would support implementing 3 as proposed, and also checking if there can simple examples how 1-2 can be achieved with a short command (just 1-3 lines perhaps?).